### PR TITLE
Src: add CapsuleGeometry

### PR DESCRIPTION
**Pull request by gsimone**
Related issue: #23579

**Preview**
 
https://rawcdn.githack.com/gsimone/three.js/feature/capsule-geometry/docs/index.html?q=caps#api/en/geometries/CapsuleGeometry
<a href="https://rawcdn.githack.com/gsimone/three.js/feature/capsule-geometry/docs/index.html?q=caps#api/en/geometries/CapsuleGeometry">image</a>

**Description**

Adds a simple CapsuleGeometry class that extends LatheGeometry.

**Some Notes**
- the naming for the arguments and the descriptions were inspired by other geometries, suggestions and improvements are welcome.
- there are combinations of radii/height that don't produce a geometry, should we try to prevent them or leave it to users?

**Checklist**

- [x] Export both Geometry and BufferGeometry
- [x] Export from Geometries.js
- [x] Add docs navigation
- [x] Add docs page w/ working example
- [x] Add unit tests

### CODE REVIEW

File: src/geometries/CapsuleGeometry.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+	/**
+	 * Creates a CapsuleGeometry.
+	 *
+	 * @param {number} radiusTop — Radius of the emisphere at the top. Optional; defaults to 1.
+	 * @param {number} radiusBottom — Radius of the emisphere at the bottom. Optional; defaults to 1.
+	 * @param {number} height — Height of the middle section. Optional; defaults to 1.
+	 * @param {number} capSegments — Number of curve segments used to build the caps. Optional; defaults to 4.
+	 * @param {number} radialSegments — Number of segmented faces around the circumference of the capsule. Optional; defaults to 8.
+	 */
```
Comment (gsimone) [inline src/geometries/CapsuleGeometry.js:14]: Should I remove these? The rest of the codebase doesn't use them so it feels irrelevant

File: src/geometries/CapsuleGeometry.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	constructor( radiusTop = 1, radiusBottom = 1, height = 1, capSegments = 4, radialSegments = 8 ) {
+
+		const R1 = radiusTop - radiusBottom;
+		// max here prevents failing when height is 0
+		const a = Math.asin( R1 / Math.max( Number.EPSILON, height ) );
```
Comment (gsimone) [inline src/geometries/CapsuleGeometry.js:19]: I added this check to avoid a void geometry. 
Problems still arise when passing certain combinations of radii and height, should we add more checks/failsafes?

**Comment by gsimone** [comment]
Not sure what's wrong with the unit tests, maybe I missed something but the action message doesn't help

File: src/geometries/CapsuleGeometry.js (hunk 1 above)
Comment (mrdoob) [reply src/geometries/CapsuleGeometry.js:14]: We can remove these yeah.

**Comment by mrdoob** [comment]
Could you remove the builds from the PR?

**Comment by gsimone** [comment]
✅  Removed them

File: src/geometries/CapsuleGeometry.js (hunk 1 above)
Comment (gsimone) [inline src/geometries/CapsuleGeometry.js:14]:
```suggestion
```

File: docs/api/en/geometries/CapsuleGeometry.html [hunk 3]
```diff
@@ -0,0 +1,74 @@
+<!DOCTYPE html>
+<html lang="en">
+	<head>
+		<meta charset="utf-8" />
+		<base href="../../../" />
+		<script src="page.js"></script>
+		<link type="text/css" rel="stylesheet" href="page.css" />
+	</head>
+	<body>
+		[page:BufferGeometry] &rarr;
```
Comment (Mugen87) [inline docs/api/en/geometries/CapsuleGeometry.html:10]: It should be:
```
[page:BufferGeometry] &rarr; [page:LatheGeometry] &rarr;
```

Comment (gsimone) [inline docs/api/en/geometries/CapsuleGeometry.html:10]:
```suggestion
		[page:BufferGeometry] &rarr; [page:LatheGeometry] &rarr;
```

Comment (gsimone) [reply docs/api/en/geometries/CapsuleGeometry.html]: Done ✅

**Comment by gsimone** [comment]
@mrdoob @WestLangley 

2 separate radii or 1? I agree that radius/length is less clumsy to use, and the change is not a big effort 🤷  Let me know which one you prefer

**Comment by Mugen87** [comment]
I vote for a single radius value.

File: src/geometries/CapsuleGeometry.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+	constructor( radius = 1, length = 1, capSegments = 4, radialSegments = 8 ) {
+
+		const R1 = 0;
+		// max here prevents failing when height is 0
+		const a = Math.asin( R1 / Math.max( Number.EPSILON, length ) );
+
+		const path = new Path();
+		path.absarc( 0, 0, radius, Math.PI * 1.5, a );
+		path.absarc( 0, length, radius, a, Math.PI * 0.5 );
```
Comment (gsimone) [inline src/geometries/CapsuleGeometry.js:14]: Since there's no diff between the two radii, this can be simplified like this

```suggestion
		const path = new Path();
		path.absarc( 0, 0, radius, Math.PI * 1.5, 0 );
		path.absarc( 0, length, radius, 0, Math.PI * 0.5 );
```

yay or nay?

**Comment by gsimone** [comment]
Removed the double radius situation, only thing missing is this suggestion https://github.com/mrdoob/three.js/pull/23586#discussion_r814983513

File: src/geometries/CapsuleGeometry.js (hunk 4 above)
Comment (mrdoob) [reply src/geometries/CapsuleGeometry.js:14]: Simpler is always better 🤓

**Comment by gsimone** [comment]
✅  simplified code accounting for single radius

**Comment by Mugen87** [comment]
You have to add the new unit test file to this section:

https://github.com/mrdoob/three.js/blob/5d81b2724cb6beb8d7663145c68a43b99e3bbbef/test/unit/three.source.unit.js#L85-L104

File: src/geometries/CapsuleGeometry.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+	constructor( radius = 1, length = 1, capSegments = 4, radialSegments = 8 ) {
+
+		const path = new Path();
+		path.absarc( 0, 0, radius, Math.PI * 1.5, 0 );
+		path.absarc( 0, length, radius, 0, Math.PI * 0.5 );
+
+		super( path.getPoints( capSegments ), radialSegments );
+
+		this.translate( 0, - length / 2, 0 );
```
Comment (mrdoob) [inline src/geometries/CapsuleGeometry.js:14]: Is there a way of doing this offset in the `Path` instead? 🤔

Comment (gsimone) [reply src/geometries/CapsuleGeometry.js]: https://github.com/mrdoob/three.js/pull/23586/commits/4ed33917f351a3c47477a5e70497df1fd93fd9e6 LGTY?

Comment (mrdoob) [reply src/geometries/CapsuleGeometry.js]: LGTM!

**Comment by mrdoob** [comment]
Thanks!

**Comment by mrdoob** [comment]
@gsimone It seems to me that `SphereGeometry` could be simplified by adopt the approach used here?

**Comment by gsimone** [comment]
@mrdoob This is what it would look like, I didn't check if it generates the same vertices though, is that something we'd care about 🤔 

https://user-images.githubusercontent.com/1862172/156137222-c4b5f342-ae51-4532-a8e0-6e214fcffcb8.mov

 Try here 👉  https://codesandbox.io/s/three-js-lathe-spheregeometry-vg4bvz?file=/src/LatheSphereGeometry.js

**Comment by gsimone** [comment]
I think it would be an interesting experiment, but maybe not worth it? I'll play with adding the angles and see how it differs from the sphere we have now. 

It could be good as long as it's not a huge breaking change (eg. UV and normals are close enough to not cause differences if people are relying on constructing spheres with the old geo)
